### PR TITLE
Profiler tutorial

### PR DESCRIPTION
**Pull request by suraj813**

### CODE REVIEW

**Comment by suraj813** [comment]
@ilia-cher pending your clarification regarding the "overhead" from `record_function` I'll update the tutorial. For now I have commented `record_function` and added a placeholder disclaimer

File: beginner_source/profiler_tutorial.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
+"""
+Profiling your PyTorch Module
+------------
+
+PyTorch includes a profiler API that is useful to identify the time and
+memory costs of various PyTorch operations in your code. It records the
+events of functions executed in C++, and exposes them to Python.
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:7]: it's not only C++, and we also have C++ version of profiler api, so let's remove this sentence.

File: beginner_source/profiler_tutorial.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+Profiler can be easily integrated in your code, and the results are
+printed in a pretty table or as a JSON trace file.
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:9]: nit: remove 'pretty'

File: beginner_source/profiler_tutorial.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+# - a linear transformation on the input, and
+# - use the transformation result to get indices on a mask tensor.
+#
+# We wrap the code for each sub-task in separate labelled context managers using
+# ``profiler.record_function("label")``. In the profiler output, the
+# aggregate performance metrics of all operations in the sub-task will
+# show up under its corresponding label.
+#
+#
+# Note that this incurs some overhead, and is best used only while
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:47]: i wouldn't single out `record_function` as a source of overhead, it's more complex (and depends on both RecordFunction and profiler)

among things that are known though to increase profiler overhead is, e.g. `with_stack=True` option

File: beginner_source/profiler_tutorial.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+# investigating the trace. Remove it if you are benchmarking runtimes.
+#
+
+class CustomLinear(torch.nn.Linear):
+    def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
+        super().__init__(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        # with profiler.record_function("LINEAR PASS"):
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:56]: let's uncomment this and below, a few user-defined `record_function`(s) should be generally ok

File: beginner_source/profiler_tutorial.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+with profiler.profile(with_stack=True, profile_memory=True) as prof:
+    out, idx = model(input, mask)
+
+
+######################################################################
+# Pretty print profiler results
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# Finally, we print the profiler results. ``profiler.key_averages``
+# aggregates the results either by input shapes or stack trace events.
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:92]: or both shapes and stack; it also groups by operator name (always)

File: beginner_source/profiler_tutorial.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+# Grouping by input shapes is useful to identify which tensor shapes
+# contribute to runtime.
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:94]: "are utilized by the model" is prob. a bit more accurate

File: beginner_source/profiler_tutorial.py [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+# Here, we use ``group_by_stack_n=5`` which aggregates runtimes by the
+# operation and its traceback (truncated to the most recent 5 events), and
+# display the events in the order they are registered. The table can also
+# be sorted by passing a ``sort_by`` argument (refer to the
+# `docs <https://pytorch.org/docs/stable/autograd.html#profiler>`__ for
+# valid sorting keys).
+#
+
+print(prof.key_averages(group_by_stack_n=5).table())
+
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:105]: we should also include the output of the profiler

File: beginner_source/profiler_tutorial.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+# operation consumes 953.67 Mb. ``mask`` is initialized with a
+# ``torch.double`` datatype. Can we reduce the memory footprint by casting
+# it to ``torch.float``?
+#
+
+model = CustomLinear(500, 10).cuda()
+input = torch.rand(128, 500).cuda()
+mask = torch.rand((500, 500, 500), dtype=torch.float).cuda()
+
+model(input, mask) #warm up
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:122]: nit: double space before #, single space after

File: beginner_source/profiler_tutorial.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+#
+
+
+######################################################################
+# Caveat
+# ^^^^^^^^^^^^^^
+#
+# You might notice introducing ``torch.nonzero`` in our module has shown a
+# new memory footprint of 2.59 Gb that wasn’t in the stacktrace when we
+# used ``np.argwhere``. This is because Profiler tracks only operations in
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:182]: let me check what's going on here;

also, more precisely, profiler tracks PyTorch tensor memory consumption, and does not track numpy tensors, but we plan to have a better memory tracking

File: beginner_source/profiler_tutorial.py [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+#
+# Thread local
+# ^^^^^^^^^^^^^^
+#
+# Profiler runs in the same thread as the operation but it will also
+# profile child operators that might run in another thread.
+# Concurrently-running profilers will be scoped to their own thread to
+# prevent mixing of results.
+#
+# Module-level, not low-level system profiling
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:200]: module -> model

tbh, we're actively working on system metrics, so maybe let's not focus too much on that or mention that extra events from the system are being added

File: beginner_source/profiler_tutorial.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+# ^^^^^^^^^^^^^^
+#
+# PyTorch profiler only reports runtime of PyTorch functions, and not at
+# the lower-level system operations.
+#
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:205]: we need some conclusion, e.g. "we considered an example and showed how to use profiler to improve time and memory consumption of the model. For more examples, please check profiler recipe/profiler [links]"

File: beginner_source/profiler_tutorial.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+# That worked, the memory footprint has halved - but while the time
+# consumed has reduced a bit (previously 283ms), it’s still too high at
+# 194ms. Turns out copying a matrix from CUDA to CPU is pretty expensive!
+# The ``aten::copy_`` operator in ``forward (11)`` copies ``mask`` to CPU
+# so that it can use the NumPy ``argwhere`` function. We could eliminate
+# the need to copy into a NumPy array if we use a ``torch`` function
+# ``nonzero()`` here instead.
+#
+
+class CustomLinearLite(torch.nn.Linear):
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:142]: let's maybe, to avoid confusion, call it `MyModule` or smth similar, so it's clear that this is a custom user model code
further, we could also explicitly create a linear module inside the ctor and not inherit from Linear

File: beginner_source/profiler_tutorial.py [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    def __init__(self, in_features: int, out_features: int, bias: bool = True) -> None:
+        super().__init__(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        # with profiler.record_function("LINEAR PASS"):
+            out = super().forward(input)
+
+        # with profiler.record_function("MASK INDICES"):
+            threshold = out.sum(axis=1).mean().item()
+            hi_idx = np.argwhere(mask.cpu().numpy() > threshold) # fix: use torch.nonzero()
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:61]: should we move it back to gpu?

**Review by ilia-cher — COMMENTED** [review]
Thanks! Left some comments.
Also cc. @ngimel

File: beginner_source/profiler_tutorial.py (hunk 3 above)
Comment (suraj813) [reply beginner_source/profiler_tutorial.py:47]: Can we add a general note about overheads then, instead of singling out `record_function`?

File: beginner_source/profiler_tutorial.py (hunk 7 above)
Comment (suraj813) [reply beginner_source/profiler_tutorial.py:105]: The output doesn't show up in the netlify preview, but will build and show when deployed

File: beginner_source/profiler_tutorial.py (hunk 9 above)
Comment (suraj813) [reply beginner_source/profiler_tutorial.py:182]: Gotcha, I'll update the note to reflect this.

File: beginner_source/profiler_tutorial.py (hunk 10 above)
Comment (suraj813) [reply beginner_source/profiler_tutorial.py:200]: Okay, I'll remove this for now as we are specifying earlier that currently only torch.Tensors are being tracked

File: beginner_source/profiler_tutorial.py (hunk 13 above)
Comment (suraj813) [reply beginner_source/profiler_tutorial.py:61]: i think it's not necessary because moving to cpu alone is pretty expensive and showcases inefficiencies of using numpy where torch functions could have worked

File: beginner_source/profiler_tutorial.py (hunk 12 above)
Comment (suraj813) [reply beginner_source/profiler_tutorial.py:142]: agreed

File: beginner_source/profiler_tutorial.py [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+# Profile the forward pass
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^
+#
+# We initialize random input and mask tensors, and the ``CustomLinear`` module.
+#
+# Before we run the profiler, we warm-up CUDA to ensure accurate
+# performance benchmarking. We wrap the forward pass of our module in the
+# ``profiler.profile`` context manager. The ``with_stack=True`` parameter appends the
+# file and line number of the operation in the trace.
+#
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:76]: pls add a warning that `with_stack=True` incurs an additional overhead

File: beginner_source/profiler_tutorial.py [hunk 15]
```diff
@@ -0,0 +1,203 @@
...
+        super(MyModule, self).__init__()
+        self.linear = nn.Linear(in_features, out_features, bias)
+
+    def forward(self, input, mask):
+        with profiler.record_function("LINEAR PASS"):
+            out = self.linear(input)
+
+        with profiler.record_function("MASK INDICES"):
+            threshold = out.sum(axis=1).mean().item()
+            hi_idx = np.argwhere(mask.cpu().numpy() > threshold)
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:61]: I think we probably want to also move hi_idx back to gpu, if the tensor was on gpu

**Comment by ilia-cher** [comment]
Looks good, thanks!
Let me take a look at what happens with memory ('caveat' section) and then it'll be good to go

File: beginner_source/profiler_tutorial.py (hunk 15 above)
Comment (ilia-cher) [reply beginner_source/profiler_tutorial.py:61]: and return pytorch tensor not a numpy array

File: beginner_source/profiler_tutorial.py [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+#
+# Here, we use ``group_by_stack_n=5`` which aggregates runtimes by the
+# operation and its traceback (truncated to the most recent 5 events), and
+# display the events in the order they are registered. The table can also
+# be sorted by passing a ``sort_by`` argument (refer to the
+# `docs <https://pytorch.org/docs/stable/autograd.html#profiler>`__ for
+# valid sorting keys).
+#
+
+print(prof.key_averages(group_by_stack_n=5).table())
```
Comment (ilia-cher) [inline beginner_source/profiler_tutorial.py:107]: when outputting stats here and in other places - let's use `table(sort_by=<metric>, row_limit=X)`

where X - is up to you, just to comfortably see the interesting parts, -1 means 'all rows'
and metric - for memory it is "cpu_memory_usage" and "cuda_memory_usage"

**Review by ilia-cher — APPROVED** [review]
thanks for adding the tutorial!
